acvm/src/pwg/block.rs

```diff
@@ -0,0 +1,164 @@
+use std::collections::{BTreeMap, HashMap};
+
+use acir::{
+    circuit::opcodes::{BlockId, MemOp},
+    native_types::Witness,
+    FieldElement,
+};
+
+use crate::{OpcodeNotSolvable, OpcodeResolutionError};
+
+use super::{
+    arithmetic::{ArithmeticSolver, GateStatus},
+    directives::insert_witness,
+    expression_to_const,
+};
+
+/// Maps a block to its emulated state
+#[derive(Default)]
+pub struct Blocks {
+    blocks: HashMap<BlockId, BlockSolver>,
+}
+
+impl Blocks {
+    pub fn solve(
+        &mut self,
+        id: BlockId,
+        trace: &[MemOp],
+        solved_witness: &mut BTreeMap<Witness, FieldElement>,
+    ) -> Result<(), OpcodeResolutionError> {
+        let solver = self.blocks.entry(id).or_default();
+        solver.solve(solved_witness, trace)
+    }
+}
+
+/// Maintains the state for solving Block opcode
+/// block_value is the value of the Block at the solved_operations step
+/// solved_operations is the number of solved elements in the block
+#[derive(Default)]
+struct BlockSolver {
+    block_value: HashMap<u32, FieldElement>,
+    solved_operations: usize,
+}
+
+impl BlockSolver {
+    fn insert_value(
+        &mut self,
+        index: u32,
+        value: FieldElement,
+    ) -> Result<(), OpcodeResolutionError> {
+        let entry = self.block_value.entry(index);
+        match entry {
+            std::collections::hash_map::Entry::Vacant(e) => {
+                e.insert(value);
+            }
+            std::collections::hash_map::Entry::Occupied(e) => {
+                if e.get() != &value {
+                    return Err(OpcodeResolutionError::UnsatisfiedConstrain);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn get_value(&self, index: u32) -> Option<FieldElement> {
+        self.block_value.get(&index).copied()
+    }
+
+    // Try to solve block operations from the trace
+    // As long as operations are resolved, we update/read from the block_value
+    // We stop when an operation cannot be resolved
+    pub fn solve(
+        &mut self,
+        initial_witness: &mut BTreeMap<Witness, FieldElement>,
+        trace: &[MemOp],
+    ) -> Result<(), OpcodeResolutionError> {
+        let missing_assignment = |witness| {
+            OpcodeResolutionError::OpcodeNotSolvable(OpcodeNotSolvable::MissingAssignment(witness))
+        };
```

Comment:
```suggestion
        let missing_assignment = |witness| {
            OpcodeResolutionError::OpcodeNotSolvable(OpcodeNotSolvable::MissingAssignment(witness.0))
        };
```

We can let this closure do the final unwrap from witness to index.

Reply:
I also added the first unwrap()